components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIKeyValidator.java

```diff
@@ -331,45 +331,45 @@ public VerbInfoDTO findMatchingVerb(MessageContext synCtx) throws ResourceNotFou
 
         String httpMethod = (String)((Axis2MessageContext) synCtx).getAxis2MessageContext().
                 getProperty(Constants.Configuration.HTTP_METHOD);
-
-        API selectedApi = null;
-        Resource selectedResource = null;
-
-        for(API api : synCtx.getConfiguration().getAPIs()){
-            if(apiContext.equals(api.getContext()) && apiVersion.equals(api.getVersion())){
-                if(log.isDebugEnabled()){
-                    log.debug("Selected API: ".concat(apiContext).concat(", Version: ").concat(apiVersion));
+        if (resourceString == null) {
+            API selectedApi = null;
+            Resource selectedResource = null;
+
+            for (API api : synCtx.getConfiguration().getAPIs()) {
+                if (apiContext.equals(api.getContext()) && apiVersion.equals(api.getVersion())) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Selected API: ".concat(apiContext).concat(", Version: ").concat(apiVersion));
+                    }
+                    selectedApi = api;
+                    break;
                 }
-                selectedApi = api;
-                break;
             }
-        }
 
-        if (selectedApi.getResources().length > 0) {
-            for (RESTDispatcher dispatcher : RESTUtils.getDispatchers()) {
-                Resource resource = dispatcher.findResource(synCtx, Arrays.asList(selectedApi.getResources()));
-                if (resource != null && Arrays.asList(resource.getMethods()).contains(httpMethod)) {
-                    selectedResource = resource;
-                    break;
+            if (selectedApi.getResources().length > 0) {
+                for (RESTDispatcher dispatcher : RESTUtils.getDispatchers()) {
+                    Resource resource = dispatcher.findResource(synCtx, Arrays.asList(selectedApi.getResources()));
+                    if (resource != null && Arrays.asList(resource.getMethods()).contains(httpMethod)) {
+                        selectedResource = resource;
+                        break;
+                    }
                 }
             }
-        }
 
-        if(selectedResource == null){
-            //No matching resource found.
-            log.error("Could not find matching resource for " + requestPath);
-            throw new ResourceNotFoundException("Could not find matching resource for " + requestPath);
-        }
+            if (selectedResource == null) {
+                //No matching resource found.
+                log.error("Could not find matching resource for " + requestPath);
```

Comment:
Much efficient and a cleaner option would be to assign the error message to variable and then use it for logging and throwing the exception.